test/distributed/fsdp/test_fsdp_grad_acc.py

```diff
@@ -0,0 +1,167 @@
+                batches.append(tuple(permute_tensor(t) for t in batch))
+            for (batch1, batch2) in itertools.combinations(batches, r=2):
+                for t1, t2 in zip(batch1, batch2):
+                    assert not torch.all(t1 == t2)
+
+            # Concatenate the batches along the given batch dimension
+            concat_batch: Tuple[torch.Tensor, ...] = tuple(
+                torch.cat(ts, dim=batch_dim) for ts in zip(*batches)
+            )
+
+            # Establish reference gradients using the concatenated batch
+            fsdp_model.zero_grad()
+            output = fsdp_model(*concat_batch)
+            ref_loss = fsdp_model.module.get_loss(concat_batch, output)
+            ref_loss.backward()
+            ref_grads = [p.grad.detach().clone() for p in fsdp_model.parameters()]
+
+            # Compute the gradients by accumulating via `no_sync()`
+            fsdp_model.zero_grad()
+            losses = []
+            with fsdp_model.no_sync():
+                for batch in batches[:-1]:  # accumulate for all but the last batch
+                    output = fsdp_model(*batch)
+                    loss = fsdp_model.module.get_loss(batch, output)
+                    loss.backward()
+                    losses.append(loss)
+            output = fsdp_model(*batches[-1])
+            loss = fsdp_model.module.get_loss(batches[-1], output)
+            loss.backward()
+            losses.append(loss)
+            acc_loss = sum(losses)
+            acc_grads = [p.grad.detach().clone() for p in fsdp_model.parameters()]
+
+            # Compare the losses and gradients
+            torch.testing.assert_allclose(ref_loss, acc_loss)
+            assert len(ref_grads) == len(acc_grads)
+            for ref_grad, acc_grad in zip(ref_grads, acc_grads):
+                assert ref_grad.device == acc_grad.device
+                assert ref_grad.size() == acc_grad.size()
+                assert ref_grad.dtype == acc_grad.dtype
+                torch.testing.assert_allclose(ref_grad, acc_grad)
+
+            # Check that the optimizer step does not error
+            optim.step()
+        finally:
+            # Restore the default dtype
+            torch.set_default_dtype(old_default_dtype)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "num_iters_to_acc",
+        [2, 4],
+    )
+    @parametrize(
+        "cpu_offload",
+        [CPUOffload(offload_params=False), CPUOffload(offload_params=True)],
+    )
+    @parametrize(
+        "backward_prefetch",
+        [BackwardPrefetch.BACKWARD_PRE, BackwardPrefetch.BACKWARD_POST, None]
+    )
+    def test_no_sync(
+        self,
+        num_iters_to_acc: int,
+        cpu_offload: CPUOffload,
+        backward_prefetch: Optional[BackwardPrefetch],
+    ):
+        """Tests the ``no_sync()`` context manager."""
+        assert num_iters_to_acc >= 2, \
+            "Accumulate for at least 2 iterations to be nontrivial"
+        self._test_no_sync(
+            batch_dim=1,
```

Comment:
The only purpose of `test_no_sync()` is to manually pass in `batch_dim=1` to `_test_no_sync()`. If this seems unnecessary, I can coalesce the two methods.